carbon-design-system/carbon
# fix(DataTable): align items top on text wrapping

Closes #12610 

Programmatically determine if a `Table` has a cell that is wrapping text content inside of it, if so apply a new `cds--data-table--top-aligned` class that will make the table align items top.

My approach here is for any given table cell, measure the textWidth inside a canvas when it is not being wrapped and compare it to the cell's width, if the rendered text is wider than the cell then the text is being wrapped. This was the only way I found to do this, definitely open to suggestion on a better implementation.

Shoutout @guidari  for pairing up with me on this 🚀

#### Changelog

**Changed**

- Call to new `setTableAlignment` function in `Table` component on first render and on resize to determine wether table should be top aligned.
- Add/Modify `Table` css to align items top and appropriate padding when necessary
- Enforce default `lg` size in `Table`

#### Testing / Reviewing

https://deploy-preview-13525--carbon-components-react.netlify.app/?path=/story/components-datatable-basic--default

1. Go to a DataTable `Playground` story
2. turn `experimentalAutoAlign` control on
3. resize window until text is wrapped, confirm items align to the top on text wrapping and that when no text is being wrapped datatable still aligns center and looks as expected. 
4. Turn `experimentalAutoAlign` control off and confirm items do not align to the top when text is wrapping
5. I know it's a lot but please test different variations and sizes 🙏🏻🙏🏻🙏🏻
Sizes: `xs`,`sm`,`md`,`lg`,`xl`
Variations:
- Expansion
- Selection (Checkbox, RadioButton)
- Actions
- With OverflowMenu
